### ISSUE TITLE
Wrong result when direct dispatch is enabled

### ISSUE THREAD

**Issue by gpopt** [greenplum-db/gpdb#14887]
Both planner and Orca generate the wrong plan and result.

## Bug Report
```sql
create table bar(a varchar);
insert into bar values('a   ');
explain select * from bar where a = 'a'::bpchar;
                                  QUERY PLAN
------------------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..431.00 rows=1 width=8)
   ->  Seq Scan on bar  (cost=0.00..431.00 rows=1 width=8)
         Filter: ((a)::bpchar = 'a'::bpchar)
 Optimizer: Pivotal Optimizer (GPORCA)
(4 rows)

select * from bar where a = 'a'::bpchar;
 a
---
(0 rows)

set optimizer=off;
explain select * from bar where a = 'a'::bpchar;
                                   QUERY PLAN
--------------------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..254.70 rows=53 width=32)
   ->  Seq Scan on bar  (cost=0.00..254.00 rows=18 width=32)
         Filter: ((a)::bpchar = 'a'::bpchar)
 Optimizer: Postgres query optimizer
(4 rows)
select * from bar where a = 'a'::bpchar;
 a
---
(0 rows)
```

Note that bpchar will remove trailing spaces of string, but the varchar will keep trailing spaces:
```sql
select length(a::bpchar) from bar;
 length
--------
      1
(1 row)

select length(a) from bar;
 length
--------
      4
(1 row)
```

When we turned off direct dispatch, we can see the correct result:
```sql
set optimizer=off;
set gp_enable_direct_dispatch=off;

explain  select * from bar where a = 'a'::bpchar;
                                   QUERY PLAN
--------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..254.70 rows=53 width=32)
   ->  Seq Scan on bar  (cost=0.00..254.00 rows=18 width=32)
         Filter: ((a)::bpchar = 'a'::bpchar)
 Optimizer: Postgres query optimizer
(4 rows)

select * from bar where a = 'a'::bpchar;
  a
------
 a
(1 row)
```
### Greenplum version or build
- planner on 6X and main branch
- Orca on main branch, Orca is correct on 6X.

We should not do direct dispatch in this case.

**Comment by kainwen** [comment]
Good catch @gpopt !

Short version:

When we find the operator is not belong to the opfamily of the hash distribution key's, we just not try to do direct dispatch. I believe ORCA should also has the same logic.

From legacy planner side, I will pair with some guys in Beijing team to fix it.

--------

TL;DR

I think we should just ignore such case just the same behavior as Postger Planner does not prune for such kind issue for partition table.

A similar example running on Postgres is as below (see it does not prune any leaf partitions).

```sql
psql (14devel)
Type "help" for help.

zlyu=# create table t (a varchar) partition by hash (a);
CREATE TABLE
zlyu=# create table t1 partition of t for values with (modulus 3,remainder 0);
CREATE TABLE
zlyu=# create table t2 partition of t for values with (modulus 3,remainder 1);
CREATE TABLE
zlyu=# create table t3 partition of t for values with (modulus 3,remainder 2);
CREATE TABLE
zlyu=#
zlyu=# explain select * from t where a = 'a'::bpchar;
                          QUERY PLAN
--------------------------------------------------------------
 Append  (cost=0.00..81.11 rows=21 width=32)
   ->  Seq Scan on t1 t_1  (cost=0.00..27.00 rows=7 width=32)
         Filter: ((a)::bpchar = 'a'::bpchar)
   ->  Seq Scan on t2 t_2  (cost=0.00..27.00 rows=7 width=32)
         Filter: ((a)::bpchar = 'a'::bpchar)
   ->  Seq Scan on t3 t_3  (cost=0.00..27.00 rows=7 width=32)
         Filter: ((a)::bpchar = 'a'::bpchar)
(7 rows)

zlyu=#
```

The Postgres Planner related pruning code is:

```c
		if (op_in_opfamily(opno, partopfamily))
		{
			get_op_opfamily_properties(opno, partopfamily, false,
									   &op_strategy, &op_lefttype,
									   &op_righttype);
		}
		else
		{
			if (part_scheme->strategy != PARTITION_STRATEGY_LIST)
				return PARTCLAUSE_NOMATCH;
```

See the logic, if it is hash partitioned, and opno is not belong to the opfamily, we cannot do any prune.
